doc/docs/config.md
# Configuration

The base Auditize configuration is done through environment variables. The following variables are available:

| Variable                               | Required or default                   | Description                                                                                                                                                                                                                                                                                                           |
| -------------------------------------- | ------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `AUDITIZE_CONFIG`                      |                                       | The path to an environment file (key-value pairs) containing the configuration. If set, the configuration will be loaded from this file instead of environment variables.                                                                                                                                             |
| `AUDITIZE_PUBLIC_URL`                  | **Required**                          | The public URL of your Auditize instance from which your users access Auditize. It will be used for instance to build the URL of the application in emails sent to users.                                                                                                                                             |
| `AUDITIZE_JWT_SIGNING_KEY`             | **Required**                          | The secret key used to sign the JWT tokens (user session cookies and access tokens). It must be a long random string. A [32 bytes long key is recommended](https://crypto.stackexchange.com/a/34866) (i.e 64 characters in hexadecimal representation), you can generate one with the command `openssl rand -hex 32`. |
| `AUDITIZE_DB_NAME`                     | `auditize`                            | The main database name.                                                                                                                                                                                                                                                                                               |
| `AUDITIZE_PG_HOST`                     | `localhost`                           | The PostgreSQL server hostname or IP address.                                                                                                                                                                                                                                                                         |
| `AUDITIZE_PG_PORT`                     | `5432`                                | The PostgreSQL server port.                                                                                                                                                                                                                                                                                           |
| `AUDITIZE_PG_USER`                     | `${USER}`                             | The PostgreSQL database user.                                                                                                                                                                                                                                                                                         |
| `AUDITIZE_PG_PASSWORD`                 |                                       | The PostgreSQL database password.                                                                                                                                                                                                                                                                                     |
| `AUDITIZE_ES_URL`                      | `http://localhost:9200`               | The Elasticsearch server URL.                                                                                                                                                                                                                                                                                         |
| `AUDITIZE_ES_USERNAME`                 |                                       | The Elasticsearch username (if authentication is enabled).                                                                                                                                                                                                                                                            |
| `AUDITIZE_ES_PASSWORD`                 |                                       | The Elasticsearch password (if authentication is enabled).                                                                                                                                                                                                                                                            |
| `AUDITIZE_ES_SSL_VERIFY`               | `true`                                | Whether to verify the SSL certificate of the Elasticsearch server.                                                                                                                                                                                                                                                    |
| `AUDITIZE_SMTP_SERVER`                 |                                       | The SMTP server used to send emails.                                                                                                                                                                                                                                                                                  |
| `AUDITIZE_SMTP_PORT`                   |                                       | The SMTP server port.                                                                                                                                                                                                                                                                                                 |
| `AUDITIZE_SMTP_USERNAME`               |                                       | The SMTP account username.                                                                                                                                                                                                                                                                                            |
| `AUDITIZE_SMTP_PASSWORD`               |                                       | The SMTP account password.                                                                                                                                                                                                                                                                                            |
| `AUDITIZE_SMTP_SENDER`                 | Defaults to `$AUDITIZE_SMTP_USERNAME` | The email address used to send emails.                                                                                                                                                                                                                                                                                |
| `AUDITIZE_LOG_EXPIRATION_SCHEDULE`     | `0 1 * * *` (every day at 1AM)        | The schedule at which expired logs are deleted.                                                                                                                                                                                                                                                                       |
| `AUDITIZE_COOKIE_SECURE`               | `false`                               | Whether the user session cookie should be [secure](https://en.wikipedia.org/wiki/Secure_cookie) (only sent over HTTPS). It is recommended to set this to `true` in production.                                                                                                                                        |
| `AUDITIZE_CORS_ALLOW_ORIGINS`          |                                       | A comma-separated list of origins allowed to make HTTP requests to Auditize.                                                                                                                                                                                                                                          |
| `AUDITIZE_USER_SESSION_TOKEN_LIFETIME` | `43200` (12 hours)                    | The lifetime of user session tokens in seconds.                                                                                                                                                                                                                                                                       |
| `AUDITIZE_ACCESS_TOKEN_LIFETIME`       | `600` (10 minutes)                    | The lifetime of access tokens in seconds.                                                                                                                                                                                                                                                                             |
| `AUDITIZE_ATTACHMENT_MAX_SIZE`         | `5242880` (5MB)                       | The maximum file size of attachments in bytes.                                                                                                                                                                                                                                                                        |
| `AUDITIZE_EXPORT_MAX_ROWS`             | `10000`                               | The maximum number of rows in exports (`0` means no limit).                                                                                                                                                                                                                                                           |


CHANGELOG.md
# 0.9.0 (2025-12-03)

- Add types (string, enum, datetime, json, boolean, integer, float) to custom fields
- Add CLI command `auditize reindex` to reindex the Elasticsearch index to the
  latest mapping version / log format

# 0.8.0 (2025-11-18)

- Improve search from actor name, resource name and tag name
- Add a global fulltext search field
- Use fulltext search for search on plain text fields (i.e. custom field values)
- Use asciifolding text analysis in Elasticsearch to make search accent-insensitive
- OpenAPI documentation improvements
- Add `POST /api/repos/{repo_id}/logs/import` endpoint (that allows setting the `id` and `saved_at` fields)
- the API now returns datetimes with milliseconds precision
- Upgrade back & front dependencies (among which Mantine 7.x => 8.3, React 18 => 19.2, React Router 6.x => 7.2)
- Fix non-working access token authentication when the token has been generated by a non-superadmin API key (used by the web component)
- Fix log entities not properly scoped by repository on read operations
- Various improvements and bug fixes

# 0.7.1 (2025-10-05)

- Fix regression in 0.7.0 that forbid user creation after last `bcrypt` upgrade

# 0.7.0 (2025-10-05)

- **Breaking changes**:
  - Switch from MongoDB to PostgreSQL (main database) and Elasticsearch (log storage)
  - Use cursor pagination for all `GET /api/repos/{repo_id}/logs/...` list endpoints such as `/api/repos/{repo_id}/logs/actions/types`
- Add CLI command `auditize empty-repo` to delete all logs from a repository
- Upgrade back & front dependencies

# 0.6.0 (2024-12-24)

- The database name can now be chosen through the `AUDITIZE_DB_NAME` env variable
- **Breaking changes**:
  - in user/apikey management, `permissions.logs.repos.readable_entities` now
    controls the list of readable entities when `permissions.logs.repos.read` is false
  - use cursor-based pagination in `GET /api/repos/{repo_id}/logs/...` endpoints
- Minor fixes
- Upgrade back & front dependencies

# 0.5.1 (2024-12-04)

- Fix log entity read restrictions not taken into account in
  `GET /api/repos/{repo_id}/logs/{log_id}/attachments/{attachment_idx}`
- Minor UI fixes

# 0.5.0 (2024-11-21)

- Improve log search performance
- Add command `auditize purge-expired-logs`
  (the expired logs purge is also managed by the `auditize scheduler` command)
- Add database migration mechanism
- Upgrade front dependencies

# 0.4.1 (2024-11-11)

- Fix regression introduced in 0.4.0 that could lead to `POST /api/repos/{repo_id}/logs`
  failures on concurrent access

# 0.4.0 (2024-11-08)

- Improve performance of `POST /api/repos/{repo_id}/logs` endpoint
- New environment variable `AUDITIZE_MONGODB_TLS` to support MongoDB TLS connections
- Add official support for Python 3.13
- Add official support for MongoDB 8
- Minor fixes and improvements
- **Breaking change**: Auditize now requires MongoDB to be configured with a Replica Set

# 0.3.0 (2024-10-24)

- Overhaul of the log filter UI
- Add favorite log filters feature
- The log search parameter selectors and the log table column selectors are now searchable
- Add an "About" entry in the user menu
- Display the last login date in the user management page
- Various minor UI improvements and bug fixes
- Upgrade backend dependencies

# 0.2.4 (2024-10-16)

- Disable Search indexing (Google, Bing, etc...) with noindex

# 0.2.3 (2024-10-11)

- New environment variable `AUDITIZE_COOKIE_SECURE` to set the `Secure` flag on the
  user session cookies (`false` by default)
- Minor improvements in the `auditize bootstrap-superadmin` command
- Minor UI bug fixes and improvements
- Rename the `AUDITIZE_BASE_URL` environment variable to `AUDITIZE_PUBLIC_URL`
- Remove the `auditize bootstrap-default-superadmin` command

# 0.2.2 (2024-10-07)

- Add a `AUDITIZE_CONFIG` environment variable to load the configuration from an environment file

# 0.2.1 (2024-10-04)

- First release on PyPI
- Add `auditize` script as a shortcut to `python -m auditize`
- Minor bug fixes and improvements in log UI
- Minor improvements in `python -m auditize` command

# 0.2.0 (2024-09-27)

- UI enhancements throughout the interface
- Complete redesign of the entity selector
- Backend and frontend dependencies update

# 0.1.0 (2024-09-04)

Initial Auditize release.


backend/src/auditize/config.py
import dataclasses
import os

from apscheduler.triggers.cron import CronTrigger
from dotenv import dotenv_values

from auditize.exceptions import (
    ConfigAlreadyInitialized,
    ConfigError,
    ConfigNotInitialized,
)

_DEFAULT_ATTACHMENT_MAX_SIZE = 1024 * 1024 * 5  # 5MB
_DEFAULT_EXPORT_MAX_ROWS = 10_000
_DEFAULT_USER_SESSION_TOKEN_LIFETIME = 60 * 60 * 12  # 12 hours
_DEFAULT_ACCESS_TOKEN_LIFETIME = 10 * 60  # 10 minutes
_DEFAULT_LOG_EXPIRATION_SCHEDULE = "0 1 * * *"


@dataclasses.dataclass
class Config:
    public_url: str
    jwt_signing_key: str
    user_session_token_lifetime: int
    access_token_lifetime: int
    attachment_max_size: int
    export_max_rows: int
    postgres_host: str
    postgres_port: int
    postgres_user: str
    postgres_password: str
    elastic_url: str
    elastic_user: str | None
    elastic_password: str | None
    elastic_ssl_verify: bool
    db_name: str
    smtp_server: str
    smtp_port: int
    smtp_username: str
    smtp_password: str
    _smtp_sender: str
    cors_allow_origins: list[str]
    cookie_secure: bool
    test_mode: bool
    online_doc: bool
    log_expiration_schedule: str

    @staticmethod
    def _validate_list(value):
        return value.split(",")

    @staticmethod
    def _validate_bool(value):
        if value == "true":
            return True
        if value == "false":
            return False
        raise ValueError(f"must be either 'true' or 'false'")

    @staticmethod
    def _validate_cron_expr(value):
        try:
            CronTrigger.from_crontab(value)
        except ValueError as exc:
            raise ValueError(f"not a valid cron expression ({exc})")
        return value

    def _validate(self):
        smtp_values_required = (
            self.smtp_server,
            self.smtp_port,
            self.smtp_username,
            self.smtp_password,
        )
        smtp_values = smtp_values_required + (self._smtp_sender,)

        if any(smtp_values) and not all(smtp_values_required):
            raise ConfigError(
                "SMTP configuration is incomplete, please provide all of the following environment variables:\n"
                "- AUDITIZE_SMTP_SERVER\n"
                "- AUDITIZE_SMTP_PORT\n"
                "- AUDITIZE_SMTP_USERNAME\n"
                "- AUDITIZE_SMTP_PASSWORD\n"
            )

        if bool(self.elastic_user) ^ bool(self.elastic_password):
            raise ConfigError(
                "ElasticSearch configuration is incomplete, please provide both AUDITIZE_ES_USER and AUDITIZE_ES_PASSWORD"
            )

    @classmethod
    def _load_from_env(cls, env):
        def required(key, validator=None):
            value = env[key]
            if validator:
                try:
                    value = validator(value)
                except ValueError as exc:
                    raise ConfigError(
                        f"Could not load configuration, invalid value {value!r} for {key!r}: {exc}"
                    )
            return value

        def optional(key, default=None, validator=None):
            try:
                return required(key, validator)
            except KeyError:
                return default

        try:
            config = cls(
                public_url=required("AUDITIZE_PUBLIC_URL"),
                jwt_signing_key=required("AUDITIZE_JWT_SIGNING_KEY"),
                user_session_token_lifetime=optional(
                    "AUDITIZE_USER_SESSION_TOKEN_LIFETIME",
                    _DEFAULT_USER_SESSION_TOKEN_LIFETIME,
                    validator=int,
                ),
                access_token_lifetime=optional(
                    "AUDITIZE_ACCESS_TOKEN_LIFETIME",
                    _DEFAULT_ACCESS_TOKEN_LIFETIME,
                    validator=int,
                ),
                attachment_max_size=optional(
                    "AUDITIZE_ATTACHMENT_MAX_SIZE",
                    default=_DEFAULT_ATTACHMENT_MAX_SIZE,
                    validator=int,
                ),
                export_max_rows=optional(
                    "AUDITIZE_EXPORT_MAX_ROWS",
                    default=_DEFAULT_EXPORT_MAX_ROWS,
                    validator=int,
                ),
                postgres_host=optional("AUDITIZE_PG_HOST", default="localhost"),
                postgres_port=optional("AUDITIZE_PG_PORT", validator=int, default=5432),
                postgres_user=optional(
                    "AUDITIZE_PG_USER", default=os.environ.get("USER", "")
                ),
                postgres_password=optional("AUDITIZE_PG_PASSWORD", default=""),
                elastic_url=optional(
                    "AUDITIZE_ES_URL", default="http://localhost:9200"
                ),
                elastic_user=optional("AUDITIZE_ES_USER", default=None),
                elastic_password=optional("AUDITIZE_ES_PASSWORD", default=None),
                elastic_ssl_verify=optional(
                    "AUDITIZE_ES_SSL_VERIFY", validator=cls._validate_bool, default=True
                ),
                db_name=optional("AUDITIZE_DB_NAME", default="auditize"),
                smtp_server=optional("AUDITIZE_SMTP_SERVER"),
                smtp_port=optional("AUDITIZE_SMTP_PORT", validator=int),
                smtp_username=optional("AUDITIZE_SMTP_USERNAME"),
                smtp_password=optional("AUDITIZE_SMTP_PASSWORD"),
                _smtp_sender=optional("AUDITIZE_SMTP_SENDER"),
                cors_allow_origins=optional(
                    "AUDITIZE_CORS_ALLOW_ORIGINS",
                    validator=cls._validate_list,
                    default=[],
                ),
                log_expiration_schedule=optional(
                    "AUDITIZE_LOG_EXPIRATION_SCHEDULE",
                    validator=cls._validate_cron_expr,
                    default=_DEFAULT_LOG_EXPIRATION_SCHEDULE,
                ),
                cookie_secure=optional(
                    "AUDITIZE_COOKIE_SECURE",
                    validator=cls._validate_bool,
                    default=False,
                ),
                ###
                # "Private" configuration
                ###
                test_mode=optional(
                    "_AUDITIZE_TEST_MODE", validator=cls._validate_bool, default=False
                ),
                online_doc=optional(
                    "_AUDITIZE_ONLINE_DOC", validator=cls._validate_bool, default=False
                ),
            )
        except KeyError as e:
            var_name = str(e)
            raise ConfigError(
                f"Could not load configuration, variable {var_name} is missing"
            )

        config._validate()

        return config

    @classmethod
    def load_from_env(cls, env=None):
        if env is None:
            env = os.environ

        if "AUDITIZE_CONFIG" in env:
            try:
                with open(env["AUDITIZE_CONFIG"]) as fh:
                    env = dotenv_values(stream=fh)
            except IOError as exc:
                raise ConfigError(
                    f"Could not load configuration from {env['AUDITIZE_CONFIG']!r}: {exc}"
                )

        return cls._load_from_env(env)

    @property
    def smtp_sender(self):
        return self._smtp_sender or self.smtp_username

    def is_smtp_enabled(self):
        return self.smtp_sender is not None

    def get_db_url(self, db_name=None):
        if db_name is None:
            db_name = self.db_name
        return "postgresql+asyncpg://%s:%s@%s:%s/%s" % (
            self.postgres_user,
            self.postgres_password,
            self.postgres_host,
            self.postgres_port,
            db_name,
        )

    def to_dict(self):
        return dataclasses.asdict(self)


_config: Config | None = None


def init_config(env=None) -> Config:
    global _config
    if _config:
        raise ConfigAlreadyInitialized()
    _config = Config.load_from_env(env)
    return _config


def get_config() -> Config:
    if not _config:
        raise ConfigNotInitialized()
    return _config


def get_or_init_config() -> Config:
    try:
        return get_config()
    except ConfigNotInitialized:
        return init_config()
